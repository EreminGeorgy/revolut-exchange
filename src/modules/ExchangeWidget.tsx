import { Card } from 'semantic-ui-react'
import { FC, useCallback, useContext, useState, useEffect, useMemo } from 'react'

import { WalletsContext } from '../contexts/WalletsContext'
import { CurrencyDataContext } from '../contexts/CurrencyDataContext'
import Header from '../components/Header'
import Rate from '../components/Rate'
import ExchangeCard from '../components/ExchangeCard'
import SwitchButton from '../components/SwitchButton'
import { InputState } from '../components/types'
import SubmitButton from '../components/SubmitButton'
import { CURRENCIES } from '../constants/currencies'

const ExchangeWidget: FC = () => {

  const { wallets, updateWallets } = useContext(WalletsContext)
  const { state, error, isFetching} = useContext(CurrencyDataContext)

  const [currencyValues, setCurrencyValues] = useState<InputState>({
    main: '',
    dependent: '',
  })

  const errorCase = useMemo(
    () => !state.isBuyMode && (wallets[state.mainCurrency] - Number(currencyValues.main) < 0)
      || state.isBuyMode && (wallets[state.dependentCurrency] - Number(currencyValues.dependent) < 0),
    [currencyValues, state, wallets]
  )

  const equalSelectorsCase = useMemo(
    () => state.mainCurrency === state.dependentCurrency,
    [state.mainCurrency, state.dependentCurrency]
  )

  const chooseValue = useCallback(
    (first: string , second: string): string => state.isBuyMode ? first : second,
    [state.isBuyMode]
  )

  const submitHandler = useCallback(
    () => {
      updateWallets({
        buy:{
          currency: chooseValue(state.mainCurrency, state.dependentCurrency) as CURRENCIES,
          amount: Number(chooseValue(currencyValues.main, currencyValues.dependent)),
        },
        sell:{
          currency: chooseValue(state.dependentCurrency, state.mainCurrency) as CURRENCIES,
          amount: Number(chooseValue(currencyValues.dependent, currencyValues.main)),
        },
      })
    },
    [updateWallets, state, currencyValues, chooseValue]
  )

  return (
    <Card className="container">
      <Header />
      <Rate data-testid="rate"/>
      <ExchangeCard 
        isMain={true}
        currencyValues={currencyValues}
        setCurrencyValues={setCurrencyValues}
        wallets={wallets}
        errorMessage={errorCase}
        data-testid="exchange-card-main"
      />
      <SwitchButton/>
      <ExchangeCard 
        isMain={false}
        currencyValues={currencyValues}
        setCurrencyValues={setCurrencyValues}
        wallets={wallets}
        errorMessage ={errorCase}
        data-testid="exchange-card-secondary"
      />
      <SubmitButton 
        disabled={!!error || isFetching || equalSelectorsCase || errorCase} 
        onClick={submitHandler} 
        currencyValues={currencyValues}
      />
    </Card>
  )
}

export default ExchangeWidget
