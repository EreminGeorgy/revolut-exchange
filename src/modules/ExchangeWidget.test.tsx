import { render } from '@testing-library/react'

import ExchangeWidget from './ExchangeWidget'

test('matches its snapshot', () => {
  const component = render(<ExchangeWidget/>)
  expect(component).toMatchSnapshot()
})