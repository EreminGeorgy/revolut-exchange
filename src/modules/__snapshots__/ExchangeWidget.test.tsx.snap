// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches its snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="ui card container"
      >
        <h1
          data-testid="header"
        >
          Sell GBP
        </h1>
        <div>
          <div
            class="ui label"
          >
            <i
              aria-hidden="true"
              class="chart line icon"
            />
            <span>
              £1.00 = $0.00
            </span>
          </div>
        </div>
        <div
          class="ui card"
        >
          <div
            class="content"
          >
            <div
              class="card-controls"
            >
              <div>
                <div
                  aria-expanded="false"
                  class="ui inline search dropdown"
                  data-testid="currency-selector-main"
                  role="combobox"
                >
                  <input
                    aria-autocomplete="list"
                    autocomplete="off"
                    class="search"
                    tabindex="0"
                    type="text"
                    value=""
                  />
                  <div
                    aria-atomic="true"
                    aria-live="polite"
                    class="divider text"
                    role="alert"
                  >
                    <i
                      class="uk flag"
                    />
                    GBP
                  </div>
                  <i
                    aria-hidden="true"
                    class="dropdown icon"
                  />
                  <div
                    class="menu transition"
                    role="listbox"
                  >
                    <div
                      aria-checked="false"
                      aria-selected="false"
                      class="item"
                      role="option"
                      style="pointer-events: all;"
                    >
                      <i
                        class="eu flag"
                      />
                      <span
                        class="text"
                      >
                        EUR
                      </span>
                    </div>
                    <div
                      aria-checked="true"
                      aria-selected="true"
                      class="active selected item"
                      role="option"
                      style="pointer-events: all;"
                    >
                      <i
                        class="uk flag"
                      />
                      <span
                        class="text"
                      >
                        GBP
                      </span>
                    </div>
                    <div
                      aria-checked="false"
                      aria-selected="false"
                      class="item"
                      role="option"
                      style="pointer-events: all;"
                    >
                      <i
                        class="us flag"
                      />
                      <span
                        class="text"
                      >
                        USD
                      </span>
                    </div>
                  </div>
                </div>
              </div>
              <div>
                <div>
                  <div
                    class="ui large transparent input currency-input"
                    data-testid="currency-input-main"
                  >
                    <input
                      placeholder="0"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
            </div>
            <div>
              <div
                class="meta"
              >
                
        Balance:
        £1,000.00
      
              </div>
            </div>
          </div>
        </div>
        <div
          class="ui clearing fitted horizontal divider"
        >
          <div>
            <button
              class="ui mini basic circular icon button"
              data-testid="switch-button"
            >
              <i
                aria-hidden="true"
                class="arrow down icon"
              />
            </button>
          </div>
        </div>
        <div
          class="ui card"
        >
          <div
            class="content"
          >
            <div
              class="card-controls"
            >
              <div>
                <div
                  aria-expanded="false"
                  class="ui inline search dropdown"
                  data-testid="currency-selector-secondary"
                  role="combobox"
                >
                  <input
                    aria-autocomplete="list"
                    autocomplete="off"
                    class="search"
                    tabindex="0"
                    type="text"
                    value=""
                  />
                  <div
                    aria-atomic="true"
                    aria-live="polite"
                    class="divider text"
                    role="alert"
                  >
                    <i
                      class="us flag"
                    />
                    USD
                  </div>
                  <i
                    aria-hidden="true"
                    class="dropdown icon"
                  />
                  <div
                    class="menu transition"
                    role="listbox"
                  >
                    <div
                      aria-checked="false"
                      aria-selected="false"
                      class="item"
                      role="option"
                      style="pointer-events: all;"
                    >
                      <i
                        class="eu flag"
                      />
                      <span
                        class="text"
                      >
                        EUR
                      </span>
                    </div>
                    <div
                      aria-checked="false"
                      aria-selected="false"
                      class="item"
                      role="option"
                      style="pointer-events: all;"
                    >
                      <i
                        class="uk flag"
                      />
                      <span
                        class="text"
                      >
                        GBP
                      </span>
                    </div>
                    <div
                      aria-checked="true"
                      aria-selected="true"
                      class="active selected item"
                      role="option"
                      style="pointer-events: all;"
                    >
                      <i
                        class="us flag"
                      />
                      <span
                        class="text"
                      >
                        USD
                      </span>
                    </div>
                  </div>
                </div>
              </div>
              <div>
                <div>
                  <div
                    class="ui large transparent input currency-input"
                    data-testid=""
                  >
                    <input
                      placeholder="0"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
            </div>
            <div>
              <div
                class="meta"
              >
                
        Balance:
        $1,000.00
      
              </div>
            </div>
          </div>
        </div>
        <button
          class="ui fluid disabled button"
          data-testid="submit"
          disabled=""
          tabindex="-1"
        >
          Insufficient funds
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="ui card container"
    >
      <h1
        data-testid="header"
      >
        Sell GBP
      </h1>
      <div>
        <div
          class="ui label"
        >
          <i
            aria-hidden="true"
            class="chart line icon"
          />
          <span>
            £1.00 = $0.00
          </span>
        </div>
      </div>
      <div
        class="ui card"
      >
        <div
          class="content"
        >
          <div
            class="card-controls"
          >
            <div>
              <div
                aria-expanded="false"
                class="ui inline search dropdown"
                data-testid="currency-selector-main"
                role="combobox"
              >
                <input
                  aria-autocomplete="list"
                  autocomplete="off"
                  class="search"
                  tabindex="0"
                  type="text"
                  value=""
                />
                <div
                  aria-atomic="true"
                  aria-live="polite"
                  class="divider text"
                  role="alert"
                >
                  <i
                    class="uk flag"
                  />
                  GBP
                </div>
                <i
                  aria-hidden="true"
                  class="dropdown icon"
                />
                <div
                  class="menu transition"
                  role="listbox"
                >
                  <div
                    aria-checked="false"
                    aria-selected="false"
                    class="item"
                    role="option"
                    style="pointer-events: all;"
                  >
                    <i
                      class="eu flag"
                    />
                    <span
                      class="text"
                    >
                      EUR
                    </span>
                  </div>
                  <div
                    aria-checked="true"
                    aria-selected="true"
                    class="active selected item"
                    role="option"
                    style="pointer-events: all;"
                  >
                    <i
                      class="uk flag"
                    />
                    <span
                      class="text"
                    >
                      GBP
                    </span>
                  </div>
                  <div
                    aria-checked="false"
                    aria-selected="false"
                    class="item"
                    role="option"
                    style="pointer-events: all;"
                  >
                    <i
                      class="us flag"
                    />
                    <span
                      class="text"
                    >
                      USD
                    </span>
                  </div>
                </div>
              </div>
            </div>
            <div>
              <div>
                <div
                  class="ui large transparent input currency-input"
                  data-testid="currency-input-main"
                >
                  <input
                    placeholder="0"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
          <div>
            <div
              class="meta"
            >
              
        Balance:
        £1,000.00
      
            </div>
          </div>
        </div>
      </div>
      <div
        class="ui clearing fitted horizontal divider"
      >
        <div>
          <button
            class="ui mini basic circular icon button"
            data-testid="switch-button"
          >
            <i
              aria-hidden="true"
              class="arrow down icon"
            />
          </button>
        </div>
      </div>
      <div
        class="ui card"
      >
        <div
          class="content"
        >
          <div
            class="card-controls"
          >
            <div>
              <div
                aria-expanded="false"
                class="ui inline search dropdown"
                data-testid="currency-selector-secondary"
                role="combobox"
              >
                <input
                  aria-autocomplete="list"
                  autocomplete="off"
                  class="search"
                  tabindex="0"
                  type="text"
                  value=""
                />
                <div
                  aria-atomic="true"
                  aria-live="polite"
                  class="divider text"
                  role="alert"
                >
                  <i
                    class="us flag"
                  />
                  USD
                </div>
                <i
                  aria-hidden="true"
                  class="dropdown icon"
                />
                <div
                  class="menu transition"
                  role="listbox"
                >
                  <div
                    aria-checked="false"
                    aria-selected="false"
                    class="item"
                    role="option"
                    style="pointer-events: all;"
                  >
                    <i
                      class="eu flag"
                    />
                    <span
                      class="text"
                    >
                      EUR
                    </span>
                  </div>
                  <div
                    aria-checked="false"
                    aria-selected="false"
                    class="item"
                    role="option"
                    style="pointer-events: all;"
                  >
                    <i
                      class="uk flag"
                    />
                    <span
                      class="text"
                    >
                      GBP
                    </span>
                  </div>
                  <div
                    aria-checked="true"
                    aria-selected="true"
                    class="active selected item"
                    role="option"
                    style="pointer-events: all;"
                  >
                    <i
                      class="us flag"
                    />
                    <span
                      class="text"
                    >
                      USD
                    </span>
                  </div>
                </div>
              </div>
            </div>
            <div>
              <div>
                <div
                  class="ui large transparent input currency-input"
                  data-testid=""
                >
                  <input
                    placeholder="0"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
          <div>
            <div
              class="meta"
            >
              
        Balance:
        $1,000.00
      
            </div>
          </div>
        </div>
      </div>
      <button
        class="ui fluid disabled button"
        data-testid="submit"
        disabled=""
        tabindex="-1"
      >
        Insufficient funds
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
